src/core/triple/triplestore.pl

```diff
@@ -231,6 +231,8 @@
  * Insert triple into transaction layer, record changed as 1 or 0
  */
 insert(G,X,Y,Z,Changed) :-
+    do_or_die(ground(Z),
+              error(instantiation_error, _)),
```

Comment:
I've started using the following for similar cases:

```prolog
write_data_version_header(Data_Version) :-
    throw(error(unexpected_argument_instantiation(write_data_version_header, Data_Version), _)).
```

Is there some error like `unexpected_argument_instantiation(insert, Z)` that would be more descriptive in identifying this problem when it appears?

Reply:
This is what prolog throws in libraries according to a standard convention. I agree it is not ideal, but we'd have to catch two different cases if we don't do it this way.

Reply:
Do we catch `instantiation_error`? I don't see where, and I don't understand why we would. It indicates a programmatic error, not a problem with the input.

Reply:
We do, and it tracks precisely the cases where a predicate is called with some uninstantiated variables which can occur in WOQL when using WOQL words which expect bound arguments in certain modes which is why we should throw the same error condition as we get from library predicates which are called in the wrong mode.

Reply:
Okay, so this is specific to `triplestore.pl` and WOQL. I thought you meant we were catching this at the route handlers.